lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -39,26 +39,93 @@ def object
               object
             end
           end
+
+          ResourceIdentifier = Struct.new(:id, :type) do
+            def self.type_for(serializer)
+              return serializer._type if serializer._type
+              if ActiveModel::Serializer.config.jsonapi_resource_type == :singular
+                serializer.object.class.model_name.singular
+              else
+                serializer.object.class.model_name.plural
+              end
+            end
+
+            def self.id_for(serializer)
+              if serializer.respond_to?(:id)
+                serializer.id
+              else
+                serializer.object.id
+              end
+            end
+
+            def self.from_serializer(serializer)
+              new(id_for(serializer), type_for(serializer))
+            end
+
+            def to_h
+              { id: id.to_s, type: type }
+            end
+          end
+
+          Resource = Struct.new(:identifier, :attributes, :relationships) do
+            def to_h
+              hash = identifier.to_h
```

Comment:
ERROR ActionController::Serialization::JsonApi::LinkedTest#test_render_collection_with_include (0.02s)
  NoMethodError:   undefined method `to_h' for #<Array:0x00000004f3dfa0>